src/client/dfuse/pil4dfs/aio.c

```diff
@@ -0,0 +1,552 @@
+		rc  = daos_eq_create(&aio_eq_list[d_aio_eq_count_g].eq);
+		if (rc)
+			goto err;
+		D_MUTEX_INIT(&aio_eq_list[d_aio_eq_count_g].lock, NULL);
+		atomic_store_relaxed(&aio_eq_list[d_aio_eq_count_g].n_op_queued, 0);
+		atomic_store_relaxed(&aio_eq_list[d_aio_eq_count_g].n_op_done, 0);
+		d_aio_eq_count_g += num_aio_eq_create;
+	}
+	aio_ctx->idx_eq = aio_ctx_total_g % d_aio_eq_count_g;
+	aio_ctx_total_g ++;
+	D_MUTEX_UNLOCK(&d_lock_aio_eqs_g);
+
+	if (d_aio_eq_count_g == 0) {
+		DS_ERROR(EBUSY, "no EQs created for AIO contexts");
+		return EBUSY;
+	}
+
+	D_MUTEX_LOCK(&aio_ctx->lock);
+	if (aio_ctx->inited == false) {
+		D_ALLOC_ARRAY(aio_ctx->ev_done_array, aio_ctx->depth);
+		if (aio_ctx->ev_done_array == NULL) {
+			D_MUTEX_UNLOCK(&aio_ctx->lock);
+			return ENOMEM;
+		}
+		/* empty queue */
+		aio_ctx->ev_done_h        = 0;
+		aio_ctx->ev_done_t        = 0;
+		aio_ctx->ev_done_array[0] = NULL;
+		aio_ctx->ev_queue_len     = 0;
+		aio_ctx->inited           = true;
+	}
+	aio_ctx->on_dfs = true;
+	D_MUTEX_UNLOCK(&aio_ctx->lock);
+
+	return 0;
+
+err:
+	D_MUTEX_UNLOCK(&d_lock_aio_eqs_g);
+	return daos_der2errno(rc);
+}
+
+static int
+aio_req_cb(void *args, daos_event_t *ev, int ret)
+{
+	D_FREE(args);
+	return 0;
+}
+
+int
+io_submit(io_context_t ctx, long nr, struct iocb *ios[])
+{
+	d_aio_ctx_t      *aio_ctx_obj = (d_aio_ctx_t *)ctx;
+	io_context_t      ctx_real    = aio_ctx_obj->ctx;
+	daos_size_t       read_size   = 0;
+	struct d_aio_ev  *ctx_ev      = NULL;
+	d_aio_req_args_t *req_args;
+	int               i, n_op_dfs, fd, io_depth;
+	int               idx_aio_eq;
+	int               rc, rc2;
+	short             op;
+	int              *fd_directed = NULL;
+
+	if (next_io_submit == NULL) {
+		next_io_submit = dlsym(RTLD_NEXT, "io_submit");
+		D_ASSERT(next_io_submit != NULL);
+	}
+	if (!d_hook_enabled)
+		goto org;
+	io_depth = aio_ctx_obj->depth;
+	if (io_depth == 0)
+		goto org;
+	if (nr > io_depth)
+		nr = io_depth;
+
+	D_ALLOC_ARRAY(fd_directed, nr);
+	if (fd_directed == NULL)
+		D_GOTO(err, rc = ENOMEM);
+
+	n_op_dfs = 0;
+	for (i = 0; i < nr; i++) {
+		fd_directed[i] = d_get_fd_redirected(ios[i]->aio_fildes);
+		if (fd_directed[i] >= FD_FILE_BASE)
+			n_op_dfs++;
+
+		op = ios[i]->aio_lio_opcode;
+		/* only support IO_CMD_PREAD and IO_CMD_PWRITE */
+		if (op != IO_CMD_PREAD && op != IO_CMD_PWRITE) {
+			DS_ERROR(EINVAL, "io_submit only supports PREAD and PWRITE for now");
+			D_GOTO(err, rc = EINVAL);
+		}
+	}
+	if (n_op_dfs == 0)
+		goto org;
+
+	if (n_op_dfs != nr) {
+		if (d_compatible_mode)
+			goto org;
+		DS_ERROR(EINVAL, "io_submit() does not support mixed non-dfs and dfs files yet in"
+			 " regular mode");
+		D_GOTO(err, rc = EINVAL);
+	}
+
+	if (!aio_ctx_obj->inited) {
+		rc = create_ev_eq_for_aio(aio_ctx_obj);
+		if (rc)
+			D_GOTO(err, rc);
+	}
+
+	for (i = 0; i < nr; i++) {
+		op         = ios[i]->aio_lio_opcode;
+		fd         = fd_directed[i] - FD_FILE_BASE;
+		idx_aio_eq = aio_ctx_obj->idx_eq;
```

Comment:
is this now the same for every io in this loop?
should it be moved out of the loop?

Reply:
Yes. "idx_aio_eq = aio_ctx_obj->idx_eq;" is same for all requests in the same context. I will move this line out of the loop. Thank you!